js/eventloop.go

```diff
@@ -0,0 +1,155 @@
+package js
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/dop251/goja"
+)
+
+// an event loop
+type eventLoop struct {
+	lock          sync.Mutex
+	queue         []func() error
+	wakeupCh      chan struct{} // maybe use sync.Cond ?
+	reservedCount int
+
+	// a map of promises which were rejected with no handler
+	// if there is something in this map at an end of a eventLoop
+	// the loop will exit with an error. This is a kin to what deno and node do
+	pendingPromiseRejections map[*goja.Promise]struct{}
+}
+
+func newEventLoop() *eventLoop {
+	return &eventLoop{
+		wakeupCh:                 make(chan struct{}, 1),
+		pendingPromiseRejections: make(map[*goja.Promise]struct{}),
+	}
+}
+
+func (e *eventLoop) wakeup() {
+	select {
+	case e.wakeupCh <- struct{}{}:
+	default:
+	}
+}
+
+// reserve "reserves" a spot on the loop, preventing it from returning/finishing. The returning function will queue it's
+// argument and wakeup the loop if needed.
+// If the eventLoop has since stopped it will not be execeted
+// This function *must* be called from within running on the event loop, but it's result can be called from anywhere
+// TODO better name
+func (e *eventLoop) reserve() func(func() error) {
+	e.lock.Lock()
+	e.reservedCount++
+	e.lock.Unlock()
+
+	return func(f func() error) {
+		e.lock.Lock()
+		e.queue = append(e.queue, f)
+		e.reservedCount--
+		e.lock.Unlock()
+		e.wakeup()
+	}
+}
+
+func (e *eventLoop) promiseRejectionTracker(p *goja.Promise, op goja.PromiseRejectionOperation) {
+	// No locking here as the goja runtime will call this synchrnously
+	// Read Notes on https://tc39.es/ecma262/#sec-host-promise-rejection-tracker
+	if op == goja.PromiseRejectionReject {
+		e.pendingPromiseRejections[p] = struct{}{}
+	} else { // goja.PromiseRejectionHandle so a promise that was previously rejected without handler now got one
+		delete(e.pendingPromiseRejections, p)
+	}
+}
+
+// initHelpers initializes some helpers around the runtime and the event loop
+// - adding setTimeout
+// - reporting (and aborting on) unhandled promise rejection
+func (e *eventLoop) initHelpers(ctx context.Context, rt *goja.Runtime) {
+	rt.SetPromiseRejectionTracker(e.promiseRejectionTracker)
+	e.addSetTimeout(ctx, rt)
+}
+
+func (e *eventLoop) getQueue() (queue []func() error, reserved bool) {
+	e.lock.Lock()
+	queue = e.queue
+	e.queue = make([]func() error, 0, len(queue))
+	reserved = e.reservedCount != 0
+	e.lock.Unlock()
+	return
+}
+
+// start will run the event loop until it's empty and there are no reserved spots
+// or a queued function returns an error. The provided function will be the first thing executed.
+// After start returns the event loop should not be reused if an error was returend
```

Comment:
Because that design turned out to be easier than anything that needed to make certain that a `Reserve` will under no circumtances be able to queue something from a previous loop. This in practice does just this, but without the need for a bunch of additonal locking and fields - you can look at earlier code (couldn't find it) for some tries at this. 

Ultimately I decided it isn't worth it especially for a first version.

Reply:
Hm thinking about this again ... maybe after waiting on the queue to be empty it's ... fine :thinking: There should really not be any issue as long as we don't have bugs :thinking:

Reply:
[52d73cb](https://github.com/grafana/k6/pull/2228/commits/52d73cb8e1139505ad0d59de1cf2d92d25d69bfb) started resuing it here